rclpy/rclpy/context.py

```diff
@@ -12,20 +12,49 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from inspect import ismethod
 import sys
 import threading
-from types import TracebackType
+from abc import abstractmethod
+from inspect import ismethod
+from types import MethodType, TracebackType
 from typing import Callable
 from typing import ContextManager
 from typing import List
 from typing import Optional
+from typing import Protocol
 from typing import Type
-import weakref
+from typing import Union
+
+from weakref import WeakMethod
+
+
+class DestroyableType(Protocol):
+    def __enter__(self) -> None:
+        ...
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_val: Optional[BaseException], exctb: Optional[TracebackType]) -> None:
+        ...
+
+    def destroy_when_not_in_use(self) -> None:
+        ...
+
+
+class ContextType(DestroyableType):
```

Comment:
I'd recommend `ContextHandle` or `InternalContext`. As a casual reader I might confuse `ContextType` and `type(Context)`